onc-healthit/lantern-back-end
# Implement val

This implements all of the validation checks that were defined in Lantern-200 and are linked to in this ticket. The validation functions were implemented as an interface so that they can be built off of in the future if we wish to validate other FHIR versions besides R4. 

To test:
- Change at least one of the endpoints in the EpicEndpointSources.json (or in the database directly) to an R4 endpoint by changing `DSTU2` in the URL to `R4`.
- Depending on how you do the previous step, running `make run` should get the new validations into the database if there are already endpoints in `fhir_endpoints`, otherwise you'll have to run `make populatedb` and then `go run cmd/main.go` in the `sendendpoints` directory.
- Highly recommend using pgAdmin to look at the validation structure. it's helpful to run something like
```
SELECT url, validation, capability_statement, tls_version, mime_types, http_response, smart_http_response FROM fhir_endpoints_info;
```
since those are the fields that are checked by the validation.
- Non-R4 endpoints should have 5 Result objects in the validation structure, R4 endpoints should have 16.
-----

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: https://oncprojectracking.healthit.gov/support/browse/LANTERN-204
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code accomplishes the tasks purpose.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge
is complete.

Create a new branch `<branch name>_gomodupdate`. Run `make update_mods` to update the `go.mod`
and `go.sum` files to point to master. Create a PR. Require one sanity check reviewer.
